File: rust-example/src/capabilities/book_stay/result.rs
use chrono::{DateTime, Utc};
use uuid::Uuid;

use super::request::Stay;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReservationConfirmed {
    pub reservation_id: Uuid,
    pub guest_id: Uuid,
    pub listing_id: Uuid,
    pub stay: Stay,
    pub guest_count: u32,
    pub confirmed_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BookingRejected {
    InvalidDateRange,
    InvalidGuestCount,
    StayStartsInPast,
    GuestNotFound,
    GuestBlocked,
    ListingNotFound,
    ListingDisabled,
    TooManyGuests,
    StayTooShort,
    StayTooLong,
    ListingUnavailable,
}

File: csharp-example/src/BookingAStay/Capabilities/BookStay/Result.cs
namespace BookingAStay.Capabilities.BookStay;

// Accepted outcome data produced by the Functional Core.
public sealed record ReservationConfirmed(
    Guid ReservationId,
    Guid GuestId,
    Guid ListingId,
    Stay Stay,
    int GuestCount,
    DateTimeOffset ConfirmedAt,
    int MaxGuestsAtConfirmation,
    int MinNightsAtConfirmation,
    int? MaxNightsAtConfirmation);

// Rejection values produced by the Functional Core.
public enum BookingRejection
{
    InvalidDateRange,
    InvalidGuestCount,
    StayStartsInPast,
    GuestNotFound,
    GuestBlocked,
    ListingNotFound,
    ListingDisabled,
    TooManyGuests,
    StayTooShort,
    StayTooLong,
    ListingUnavailable,
}

// Output value returned by the BookStay Functional Core.
public abstract record BookingOutcome
{
    private BookingOutcome()
    {
    }

    public sealed record Confirmed(ReservationConfirmed Reservation) : BookingOutcome;

    public sealed record Rejected(BookingRejection Reason) : BookingOutcome;
}
